src/emscripten_main.cpp

```diff
@@ -0,0 +1,95 @@
+/* Copyright (C) 2016, Nikolai Wuttke. All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Welcome to the RigelEngine code base! If you are looking for the place in
+// the code where everything starts, that would be in main.cpp. This file contains the
+// main() function entry point for running RigelEngine in webassembly using emscripten.
+// During all non-webassembly compilations, this file is ignored.
+
+#include "base/warnings.hpp"
+
+RIGEL_DISABLE_WARNINGS
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+RIGEL_RESTORE_WARNINGS
+
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include "game_main.hpp"
+#include "game_main.ipp"
```

Comment:
`.ipp` files are normally meant to be an implementation detail. Instead of including it here, it should be included at the end of `game_main.hpp`. But in https://github.com/lethal-guitar/RigelEngine/pull/561, I've added some functions that make it possible to create the game and call `runOneFrame` without needing to include the `.ipp` at all (it's the last commit).

Reply:
ok, I always wondered the reasoning behind ipp files. Are they part of PIMPLE kind of pattern ?

Reply:
You could use them for the pimpl-pattern, but it's a more general convention. The general use case is when you have inline functions or templates. If you have "normal" functions and classes, you have declarations in the header file, and implementation in the `.cpp` file. The interface is very readable and clean this way, because the header only contains the most important details.

But for templates and inline functions, you need to put your code into the header file. That can hurt readability. By moving the implementations into a separate `.ipp` file, and then including that file at the end of the header file, you can keep the interface nice and readable, while still making all the code visible to the compiler thanks to including the `.ipp` in the header.

That's the typical convention. I sometimes use `.ipp` for slightly different purposes, though, and this is also the case for `game_main.ipp`. In this case, I wanted to keep the declaration of the `Game` class separate from the implementation in the `.cpp` file, but I also didn't want to expose the `Game` class to the code that includes `game_main.hpp`, mainly for better compilation times. If `game_main.hpp` would include `game_main.ipp`, then `main.cpp` would transitively include all the headers included by `game_main.ipp`, and that would increase compile time for `main.cpp`.

The functions I introduced for creating a `Game` and calling `runOneFrame` use a forward declaration of the `Game` class, and thus don't need the declaration of `Game` to be available. It's a very similar technique to the pimpl-pattern.